File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/array/DeleteFirstTest.java
package com.fincity.nocode.kirun.engine.function.system.array;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonArray;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;

import reactor.test.StepVerifier;

class DeleteFirstTest {

	@Test
	void test() {
		var arr = new JsonArray();
		arr.add('c');
		arr.add('p');
		arr.add('i');
		arr.add('e');

		DeleteFirst ad = new DeleteFirst();

		var arr1 = new JsonArray();

		arr1.add('p');
		arr1.add('i');
		arr1.add('e');

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", arr))
				.setContext(Map.of()).setSteps(Map.of());

		StepVerifier.create(ad.execute(fep))
				.expectNextMatches(result -> result.next().getResult().get("source").equals(arr1));
	}

	@Test
	void test2() {
		var arr = new JsonArray();

		DeleteFirst ad = new DeleteFirst();

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", arr))
				.setContext(Map.of()).setSteps(Map.of());

		StepVerifier.create(ad.execute(fep)).expectError(KIRuntimeException.class).verify();
	}

	@Test
	void test5() {

		var array1 = new JsonArray();

		array1.add("test");
		array1.add("Driven");
		array1.add("developement");
		array1.add("I");
		array1.add("am");

		var js1 = new JsonObject();

		js1.addProperty("boolean", false);
		js1.add("array", array1);
		js1.addProperty("char", 'o');

		var js2 = new JsonObject();

		js2.addProperty("boolean", false);
		js2.add("array", array1);
		js2.addProperty("char", "asd");

		var js3 = new JsonObject();
		js3.add("array", array1);

		var js4 = new JsonObject();

		js4.addProperty("boolean", false);
		js4.add("array", array1);
		js4.addProperty("char", 'o');

		var arr = new JsonArray();

		arr.add(js1);
		arr.add(js2);
		arr.add(js3);
		arr.add(js4);
		arr.add(js1);

		var res = new JsonArray();
		res.add(js2);
		res.add(js3);
		res.add(js4);
		res.add(js1);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", arr))
				.setContext(Map.of()).setSteps(Map.of());

		DeleteFirst del = new DeleteFirst();
		StepVerifier.create(del.execute(fep))
				.expectNextMatches(result -> result.next().getResult().get("source").equals(res));
	}

	@Test
	void test4() {

		var array1 = new JsonArray();

		array1.add("test");
		array1.add("Driven");
		array1.add("developement");
		array1.add("I");
		array1.add("am");
		array1.add("using");
		array1.add("eclipse");
		array1.add("I");
		array1.add("to");
		array1.add("test");
		array1.add("the");
		array1.add("changes");
		array1.add("with");
		array1.add("test");
		array1.add("Driven");
		array1.add("developement");

		var array2 = new JsonArray();

		array2.add("test");
		array2.add("Driven");
		array2.add("developement");
		array2.add("I");
		array2.add("am");
		array2.add("using");
		array2.add("eclipse");
		array2.add("I");
		array2.add("to");
		array2.add("test");
		array2.add("the");
		array2.add("changes");
		array2.add("with");

		var array3 = new JsonArray();

		array3.add("test");
		array3.add("Driven");
		array3.add("developement");
		array3.add("I");
		array3.add("am");
		array3.add("using");
		array3.add("eclipse");
		array3.add("I");
		array3.add("to");
		array3.add("test");
		array3.add("the");
		array3.add("changes");
		array3.add("with");
		array3.add("test");
		array3.add("Driven");
		array3.add("developement");

		var array4 = new JsonArray();

		array4.add("test");
		array4.add("Driven");
		array4.add("developement");
		array4.add("I");
		array4.add("am");
		array4.add("using");
		array4.add("eclipse");
		array4.add("I");
		array4.add("to");

		var arr = new JsonArray();
		arr.add(array1);
		arr.add(array3);
		arr.add(array2);
		arr.add(array4);
		arr.add(array1);

		var res = new JsonArray();
		res.add(array3);
		res.add(array2);
		res.add(array4);
		res.add(array1);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", arr))
				.setContext(Map.of()).setSteps(Map.of());

		DeleteFirst freq = new DeleteFirst();

		StepVerifier.create(freq.execute(fep))
				.expectNextMatches(result -> result.next().getResult().get("source").equals(res));
	}

	@Test
	void test6() {
		// var arr = new JsonArray();

		DeleteFirst ad = new DeleteFirst();

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", JsonNull.INSTANCE)).setContext(Map.of()).setSteps(Map.of());

		StepVerifier.create(ad.execute(fep)).expectError(KIRuntimeException.class).verify();
	}
}

File: kirun-js/__tests__/engine/function/system/array/DeleteFirstTest.ts
import { DeleteFirst } from '../../../../../src/engine/function/system/array/DeleteFirst';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';

test('DeleteFirst Test 1', () => {
    let delet: DeleteFirst = new DeleteFirst();

    let source: any[] = [12, 14, 15, 9];

    let temp: any[] = [14, 15, 9];

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters()
        .setArguments(new Map([[DeleteFirst.PARAMETER_ARRAY_SOURCE.getParameterName(), source]]))
        .setOutput(new Map([]))
        .setContext(new Map([]));

    delet.execute(fep);

    expect(source).toStrictEqual(temp);
});

test('DeleteFirst Test 2', () => {
    let delet: DeleteFirst = new DeleteFirst();

    let source: any[] = ['c', 'p', 'i', 'e'];

    let temp: any[] = ['p', 'i', 'e'];

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters()
        .setArguments(new Map([[DeleteFirst.PARAMETER_ARRAY_SOURCE.getParameterName(), source]]))
        .setOutput(new Map([]))
        .setContext(new Map([]));

    delet.execute(fep);

    expect(source).toStrictEqual(temp);
});

test('DeleteFirst Test 3', () => {
    let delet: DeleteFirst = new DeleteFirst();

    let source: any[] = [];

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters()
        .setArguments(new Map([[DeleteFirst.PARAMETER_ARRAY_SOURCE.getParameterName(), source]]))
        .setOutput(new Map([]))
        .setContext(new Map([]));

    expect(() => delet.execute(fep)).toThrow();
});

test('DeleteFirst Test 4', () => {
    let delet: DeleteFirst = new DeleteFirst();

    var array1 = ['test', 'Driven', 'developement', 'I', 'am'];

    var js1 = {
        boolean: false,
        array: array1,
        char: 'o',
    };

    var js2 = {
        boolean: false,
        array: array1,
        char: 'asd',
    };

    var js3 = {
        array: array1,
    };

    var js4 = {
        boolean: false,
        array: array1,
        char: 'ocbfr',
    };

    var arr: any[] = [js1, js2, js3, js4, js1];

    var res: any[] = [js2, js3, js4, js1];

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters()
        .setArguments(new Map([[DeleteFirst.PARAMETER_ARRAY_SOURCE.getParameterName(), arr]]))
        .setOutput(new Map([]))
        .setContext(new Map([]));

    delet.execute(fep);

    expect(arr).toStrictEqual(res);
});

test('DeleteFirst Test 5', () => {
    let delet: DeleteFirst = new DeleteFirst();

    let source = null;

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters()
        .setArguments(new Map([[DeleteFirst.PARAMETER_ARRAY_SOURCE.getParameterName(), source]]))
        .setOutput(new Map([]))
        .setContext(new Map([]));

    expect(() => delet.execute(fep)).toThrow();
});
